File: ruby/lib/NumericBadConsequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "TreasureKind"
require_relative "BadConsequence"
module NapakalakiGame
@@MAXTREASURES = 10

class NumericBadConsequence < BadConsequence
  attr_reader :nVisibleTreasures, :nHiddenTreasures
  
  def initialize(aText, someLevels, someVisibleTreasures, someHiddenTreasures)
    super(aText, someLevels, false)
    @nVisibleTreasures = someVisibleTreasures
    @nHiddenTreasures = someHiddenTreasures
  end
public_class_method :new

def isEmpty
		return(@nVisibleTreasures + @nHiddenTreasures == 0)
end

def substractVisibleTreasure( t = nil )
  if(nVisibleTreasures > 0) then
    nVisibleTreasures -= 1
  end
end

def substractHiddenTreasure (t = nil)
  if(nHiddenTreasures > 0) then
    nHiddenTreasures -= 1
  end
end

def adjustToFitTreasureLists(v,h)
  if(@nVisibleTreasures > v.length)
    visiblesBc = v.length
  else
    visiblesBc = @nVisiblesTreasures
  end
  
  if(@nHiddenTreasures > h.length)
    hiddenBc = h.length
  else
    hiddenBc = @nHiddenTreasures
  end
  
  bc = NumericBadConsequence.new(@text, @levels, visibleBc, hiddenBc)
  return bc
end

#to_s
end
end

File: Java/src/Napakalaki/NumericBadConsequence.java
package Napakalaki;
import java.util.*;

public class NumericBadConsequence extends BadConsequence {
    
    protected int nVisibleTreasures;
    protected int nHiddenTreasures;
    
    public NumericBadConsequence( String test, int levels, int v, int h ) {
        super( test, levels );
        nVisibleTreasures = v;
        nHiddenTreasures = h;
    }

    public NumericBadConsequence() {
        nVisibleTreasures = 0;
        nHiddenTreasures = 0;
    }
		
		public int getNVisibleTreasures() {
			return nVisibleTreasures;
		}
		
		public int getNHiddenTreasures() {
			return nHiddenTreasures;
		}
    
    @Override
    public boolean isEmpty() {
        return ( nHiddenTreasures == 0 && nVisibleTreasures == 0 );
    }
    
    @Override
    public void substractVisibleTreasure( Treasure t ) {
			if ( nVisibleTreasures > 0 )
				nVisibleTreasures--;
    }
    
    @Override
    public void substractHiddenTreasure( Treasure t ) {
			if ( nHiddenTreasures > 0 )
				nHiddenTreasures--;
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists( ArrayList<Treasure> v , ArrayList<Treasure> h )
    {
        BadConsequence local_bc = null;
        
        if(nVisibleTreasures > 0 || nHiddenTreasures > 0){
            int local_nVisibleTreasures;
            int local_nHiddenTreasures;
            if(nVisibleTreasures < v.size())
                local_nVisibleTreasures = nVisibleTreasures;
            else
                local_nVisibleTreasures = v.size();
            
            if(nHiddenTreasures < h.size())
                local_nHiddenTreasures = nHiddenTreasures;
            else
                local_nHiddenTreasures = h.size();
            
            local_bc = new NumericBadConsequence("",0 , local_nVisibleTreasures, local_nHiddenTreasures);
        }
        
        return local_bc;
    }
    
    @Override
    public String toString(){
        return this.text;
    }
}
